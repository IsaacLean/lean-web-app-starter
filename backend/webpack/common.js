/**
 * This is the webpack configuration that is used for all modes.
 */
const path = require('path')
const setupNodeExternals = require('ljas-webpack/setupNodeExternals')
const { compileTS } = require('ljas-webpack')
const { merge } = require('webpack-merge')

module.exports = merge([
  {
    entry: './src/main.ts',
  },

  compileTS({ include: path.resolve(__dirname, '../src') }),

  setupNodeExternals(),
])
