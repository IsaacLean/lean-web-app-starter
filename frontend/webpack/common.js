/**
 * This is the webpack configuration that is used for all modes.
 */
const buildHTML = require('ljas-webpack/buildHTML')
const path = require('path')
const { compileReact } = require('ljas-webpack')
const { merge } = require('webpack-merge')

module.exports = merge([
  {
    entry: './src/main.tsx',
  },

  buildHTML('Lean JavaScript Application Starter', 'src/index.html'),

  compileReact(path.resolve(__dirname, '../src')),
])
