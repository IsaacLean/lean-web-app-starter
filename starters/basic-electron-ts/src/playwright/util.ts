import {
  ElectronApplication,
  Page,
  _electron as electron,
} from "@playwright/test";

/**
 * Display console API method calls from the Electron application in the Playwright console.
 * @param electronApp Electron application representation
 */
export const enableConsole = (electronApp: ElectronApplication) => {
  electronApp.on("console", (data) => console.log(data));
};

/**
 * Get the operating system that the browser is running in.
 * @param page Electron Page instance
 * @returns A promise that will resolve to the Browser's operating system type, or null if
 *     it could not be determined
 */
export const getBrowserOs = async (page: Page) => {
  const userAgent = await page.evaluate(async () => window.navigator.userAgent);

  if (userAgent.includes("Windows")) {
    return "win";
  } else if (userAgent.includes("Mac")) {
    return "mac";
  } else if (userAgent.includes("X11")) {
    return "unix";
  } else if (userAgent.includes("Linux")) {
    return "linux";
  }

  return null;
};

/**
 * Launch the Electron application with either the development or production build.
 * By default the PLAYWRIGHT_BUILD_TYPE environment variable will determine which build will be used,
 * but you can override this through some of the available parameters.
 * @param arg Parameter object
 * @param arg.buildPath Manually set the build path. Overrides buildType parameter and PLAYWRIGHT_BUILD_TYPE environment variable.
 * @param arg.buildType Determines if the development or the production build is used for testing. Overrides PLAYWRIGHT_BUILD_TYPE environment variable.
 * @param arg.launchOptions Options to pass to Electron.launch method
 * @returns A promise that will resolve to an Electron application representation
 */
export const launchElectron = ({
  buildPath,
  buildType,
  launchOptions,
}: {
  buildPath?: string;
  buildType?: "development" | "production";
  launchOptions?: Parameters<typeof electron.launch>[0];
} = {}) => {
  let arg = ".";
  if (buildPath) {
    arg = buildPath;
  } else if (
    buildType === "production" ||
    process.env.PLAYWRIGHT_BUILD_TYPE === "production"
  ) {
    arg = "./build/production/main/main.js";
  }

  return electron.launch({
    args: [arg],
    ...launchOptions,
  });
};
