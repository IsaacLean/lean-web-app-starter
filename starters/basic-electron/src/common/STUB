STUB: All code shared across main, preload, and renderer directories goes in here.
