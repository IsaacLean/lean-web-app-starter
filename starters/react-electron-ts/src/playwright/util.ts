import {
    ElectronApplication,
    Page,
    _electron as electron,
} from '@playwright/test'

/**
 * Display console API method calls from the Electron application in the Playwright console.
 * @param electronApp Electron application representation
 */
export const enableConsole = (electronApp: ElectronApplication) => {
    electronApp.on('console', (data) => console.log(data))
}

/**
 * Get the operating system that the browser is running in.
 * @param page Electron Page instance
 * @returns A promise that will resolve to the Browser's operating system type, or null if
 *     it could not be determined
 */
export const getBrowserOs = async (page: Page) => {
    const userAgent = await page.evaluate(
        async () => window.navigator.userAgent,
    )

    if (userAgent.includes('Windows')) {
        return 'win'
    } else if (userAgent.includes('Mac')) {
        return 'mac'
    } else if (userAgent.includes('X11')) {
        return 'unix'
    } else if (userAgent.includes('Linux')) {
        return 'linux'
    }

    return null
}

/**
 * Launch the Electron application with either the development or production build.
 * @param arg Parameter object
 * @param arg.buildType Determines if the development or the production build is used for testing
 * @param arg.colorScheme Emulates 'prefers-colors-scheme' media feature, supported values are 'light', 'dark', 'no-preference'. See page.emulateMedia([options]) for more details. Passing null resets emulation to system defaults. Defaults to 'light'.
 * @returns A promise that will resolve to an Electron application representation
 */
export const launchElectron = ({
    buildType,
    colorScheme,
}: {
    buildType?: 'development' | 'production'
    colorScheme?: null | 'light' | 'dark' | 'no-preference'
} = {}) => {
    const arg =
        buildType === 'production' ? './build/production/main/main.js' : '.'

    return electron.launch({
        args: [arg],
        colorScheme,
    })
}
