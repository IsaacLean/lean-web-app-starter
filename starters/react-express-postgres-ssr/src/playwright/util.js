/**
 * Load the frontend application with either the development or end-to-end host.
 * By default the PLAYWRIGHT_HOST environment variable will be used. If the E2E environment
 * variable is true, then the PLAYWRIGHT_HOST_E2E environment variable will be used instead.
 * You can override these environment variables through some of the available parameters.
 * @param page Playwright Page instance
 * @param arg Parameter object
 * @param arg.gotoOptions Options to pass to Page.goto method
 * @param arg.serverType Determines if the development or end-to-end host is used for testing. Overrides E2E environment variable.
 * @param arg.url Manually set the URL. Overrides all behaviors caused by environment variables.
 * @returns A promise that will resolve to the main resource response
 */
export const loadApp = (page, { gotoOptions, serverType, url } = {}) => {
    let u = process.env.E2E
        ? process.env.PLAYWRIGHT_HOST_E2E
        : process.env.PLAYWRIGHT_HOST
    if (url) {
        u = url
    } else if (process.env.PLAYWRIGHT_HOST_E2E && serverType === 'e2e') {
        u = process.env.PLAYWRIGHT_HOST_E2E
    }

    if (!u) {
        throw new Error('Host was not set')
    }

    return page.goto(u, gotoOptions)
}
