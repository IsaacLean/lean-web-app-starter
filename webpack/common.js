/**
 * This is the webpack configuration that is used for all modes.
 */
const { merge } = require('webpack-merge')
const buildHTML = require('../ljas-webpack/buildHTML')

module.exports = merge([
  {
    entry: './src/main.tsx',
  },

  buildHTML(),
])
