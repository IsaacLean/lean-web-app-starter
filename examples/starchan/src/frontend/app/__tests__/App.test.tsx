/**
 * @jest-environment jsdom
 */
import {
    render,
    screen,
    waitForElementToBeRemoved,
} from '@testing-library/react'

import { TestApp, setupDefaultMsw } from '../../../common/util/test'
import { buildStore } from '../../common/redux'

setupDefaultMsw()

test('App component matches snapshot for basic render', () => {
    const store = buildStore()

    const { asFragment } = render(
        <TestApp initialEntries={['/']} store={store} />
    )

    expect(asFragment()).toMatchSnapshot()
})

test('App component matches diff snapshot transitioning from loading render to complete render', async () => {
    expect.assertions(1)

    const store = buildStore()

    const { asFragment } = render(
        <TestApp initialEntries={['/']} store={store} />
    )

    const loadingRender = asFragment()

    await waitForElementToBeRemoved(() => screen.getByText(/loading.../i))

    const completeRender = asFragment()

    expect(loadingRender).toMatchDiffSnapshot(completeRender)
})
