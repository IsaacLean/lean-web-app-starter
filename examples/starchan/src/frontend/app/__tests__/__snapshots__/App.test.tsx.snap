// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component matches diff snapshot transitioning from loading render to complete render 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -5,15 +5,50 @@
      >
        *chan
      </h1>
    </header>
    <hr />
-   <div
-     class="loading"
+   <main>
+     <section
+       class="post-form"
+     >
+       <span
+         class="post-form__txt"
+       >
+         [
+         <button
+           class="post-form__btn"
+         >
+           Start a New Thread
+         </button>
+         ]
+       </span>
+       <noscript />
+     </section>
+     <hr />
+     <p>
+       Currently no threads exist. Why don't you create the first one?
+     </p>
+   </main>
+   <nav>
+     <ol
+       class="page-select"
+       data-testid="page-select"
+     >
+       <li
+         class="page-select__page page-select__page--active"
+       >
+         [
+         <a
+           href="/"
          >
-     Loading...
-   </div>
+           1
+         </a>
+         ]
+       </li>
+     </ol>
+   </nav>
    <footer
      class="footer"
      id="bottom"
    >
      <p>"
`;

exports[`App component matches snapshot for basic render 1`] = `
<DocumentFragment>
  <header>
    <h1
      id="title"
    >
      *chan
    </h1>
  </header>
  <hr />
  <div
    class="loading"
  >
    Loading...
  </div>
  <footer
    class="footer"
    id="bottom"
  >
    <p>
      *chan (pronounced as starchan) is a 
      <a
        href="https://en.wikipedia.org/wiki/Textboard"
      >
        textboard
      </a>
       that demonstrates one of 
      <a
        href="#"
      >
        Lean JS App Starter's full-stack project setups
      </a>
      . It showcases an Express server that acts as a REST API and a React server-side renderer. These work in tandem with a React frontend.
    </p>
    <nav>
      <ul
        class="footer__links"
      >
        <li>
          <a
            href="#"
          >
            About
          </a>
        </li>
        <li>
          <a
            href="#"
          >
            *chan GitHub
          </a>
        </li>
        <li>
          <a
            href="https://github.com/mattlean/lean-js-app-starter"
          >
            Lean JavaScript Application Starter
          </a>
        </li>
      </ul>
    </nav>
  </footer>
</DocumentFragment>
`;
